mlir/test/Quantum/ChainedSelfInverseTest.mlir

```diff
@@ -37,3 +37,217 @@ func.func @test_chained_self_inverse(%arg: !quantum.bit) -> !quantum.bit {
+    %out_qubits:2 = quantum.custom "CZ"() %1, %2 : !quantum.bit, !quantum.bit
+    %out_qubits_0:2 = quantum.custom "CZ"() %out_qubits#1, %out_qubits#0 : !quantum.bit, !quantum.bit
+    // CHECK: return %out_qubits_0#0, %out_qubits_0#1 : !quantum.bit, !quantum.bit
+    return %out_qubits_0#0, %out_qubits_0#1 : !quantum.bit, !quantum.bit
+}
+
+// -----
+
+// Test for chained SWAP with wrong order
+// CHECK-LABEL: test_chained_self_inverse
+func.func @test_chained_self_inverse() -> (!quantum.bit, !quantum.bit) {
+    %0 = quantum.alloc( 2) : !quantum.reg
+    %1 = quantum.extract %0[ 0] : !quantum.reg -> !quantum.bit
+    %2 = quantum.extract %0[ 1] : !quantum.reg -> !quantum.bit
+    // CHECK: quantum.custom "SWAP"() %1, %2 : !quantum.bit, !quantum.bit
+    // CHECK-NEXT: quantum.custom "SWAP"() %out_qubits#1, %out_qubits#0 : !quantum.bit, !quantum.bit
+    %out_qubits:2 = quantum.custom "SWAP"() %1, %2 : !quantum.bit, !quantum.bit
+    %out_qubits_0:2 = quantum.custom "SWAP"() %out_qubits#1, %out_qubits#0 : !quantum.bit, !quantum.bit
+    // CHECK: return %out_qubits_0#0, %out_qubits_0#1 : !quantum.bit, !quantum.bit
+    return %out_qubits_0#0, %out_qubits_0#1 : !quantum.bit, !quantum.bit
+}
+
+// -----
+
+// Test for chained Toffoli with wrong order
+// CHECK-LABEL: test_chained_self_inverse
+func.func @test_chained_self_inverse() -> (!quantum.bit, !quantum.bit, !quantum.bit) {
+    %0 = quantum.alloc(3) : !quantum.reg
+    %1 = quantum.extract %0[0] : !quantum.reg -> !quantum.bit
+    %2 = quantum.extract %0[1] : !quantum.reg -> !quantum.bit
+    %3 = quantum.extract %0[2] : !quantum.reg -> !quantum.bit
+    // CHECK: quantum.custom "Toffoli"() %1, %2, %3 : !quantum.bit, !quantum.bit, !quantum.bit
+    // CHECK-NEXT: quantum.custom "Toffoli"() %out_qubits#1, %out_qubits#2, %out_qubits#0 : !quantum.bit, !quantum.bit, !quantum.bit
+    %out_qubits:3 = quantum.custom "Toffoli"() %1, %2, %3: !quantum.bit, !quantum.bit, !quantum.bit
+    %out_qubits_0:3 = quantum.custom "Toffoli"() %out_qubits#1, %out_qubits#2, %out_qubits#0 : !quantum.bit, !quantum.bit, !quantum.bit
+    // CHECK: return %out_qubits_0#0, %out_qubits_0#1, %out_qubits_0#2 : !quantum.bit, !quantum.bit, !quantum.bit
+    return %out_qubits_0#0, %out_qubits_0#1, %out_qubits_0#2 : !quantum.bit, !quantum.bit, !quantum.bit
+}
+
+// -----
+
+// test nested self-inverse Gates with different names
+// CHECK-LABEL: test_chained_self_inverse
+func.func @test_chained_self_inverse() -> !quantum.bit {
+    // CHECK: quantum.alloc
+    // CHECK: quantum.extract
+    %0 = quantum.alloc( 1) : !quantum.reg
+    %1 = quantum.extract %0[ 0] : !quantum.reg -> !quantum.bit
+    // CHECK-NOT: quantum.custom
+    %2 = quantum.custom "Hadamard"() %1 : !quantum.bit
+    %3 = quantum.custom "PauliX"() %2 : !quantum.bit
+    %4 = quantum.custom "PauliX"() %3 : !quantum.bit
+    %5 = quantum.custom "Hadamard"() %4 : !quantum.bit
+    return %5 : !quantum.bit
+}
```

Comment:
Note that we end all our files with a newline.

Reply:
good catch! Wondering if there is anything like clang-format to catch these in IR files 🤔

Reply:
You can set preferences in your editor. I think you use vscode? https://neilsonwebdesign.com/vs-code-force-newline-at-end-of-file/#:~:text=vscode%2Fsettings.,in%20the%20project%20when%20saving